### PR TITLE
Removing access to invoice if the members don't have billing permission

### PR DESCRIPTION
**Pull request by HemaHG**
**What this PR does / why we need it**:

Removing access to invoice if the members don't have permission

**Which issue(s) this PR fixes** 
https://issues.redhat.com/browse/THREESCALE-1609, Members that don't have billing permissions have read/write access to invoices

**Verification steps** 
Verification steps is mentioned in the JIRA ticket.

### CODE REVIEW

File: app/views/buyers/accounts/_menu.html.slim [hunk 1]
```diff
@@ -41,10 +41,16 @@ ul#subsubmenu
...
+      - if can?(:manage, Invoice)
+        = link_to_unless_current_styled pluralize(number_to_human(@account.invoices.count), 'Invoice'),
+          admin_buyers_account_invoices_path(@account),
+          title: "invoices",
+          class: "SubSubMenuItem-link
+      - else
+        = link_to_unless_current_styled pluralize(number_to_human(@account.invoices.count), 'Invoice'),
+          admin_buyers_account_path(@account),
+          title: "invoices",
+          class: "SubSubMenuItem-link"
```
Comment (Martouta) [inline app/views/buyers/accounts/_menu.html.slim:53]: @thomasmaas should say what should appear here

File: app/controllers/buyers/invoices_controller.rb [hunk 2]
```diff
@@ -46,6 +47,10 @@ def update
   end
 
   private
+  def authorize_finance
+    authorize! :read, :finance
```
Comment (Martouta) [inline app/controllers/buyers/invoices_controller.rb:51]: It shouldn't be always 'read'. One can have permission to read but not to update for example.

File: test/integration/buyers/accounts_controller_test.rb [hunk 3]
```diff
@@ -91,6 +91,13 @@ def test_show
       assert_match 'Alaska Application App', response.body
     end
 
+    test "memeber user without billing permission cannot view invoice" do
+      @provider.settings.allow_finance!
+      buyer = FactoryBot.create(:buyer_account, provider_account: provider)
+      get admin_buyers_account_invoices_path(buyer)
+      assert_response :forbidden
+    end
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:99]: Good, but this should test also:
1. When the member user has finance permission, it should work!
2. You shouldn't test only 'read/show' but also the rest of actions

**Comment by Martouta** [comment]
I see the Jira issue also says:

> The request is to disable the ability to access to the invoice. Currently the member can also add new line items in the invoice.

Is this intended to be fixed in this same PR or in another PR? IMO it is better in another PR, and you may be thinking the same, but I wanna hear/read what you think 😛

File: test/integration/buyers/accounts_controller_test.rb [hunk 4]
```diff
@@ -91,6 +91,42 @@ def test_show
       assert_match 'Alaska Application App', response.body
     end
 
+    test "memeber user without billing permission cannot manage invoice" do
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:94]: 👀 

```suggestion
    test "member user without billing permission cannot manage invoice" do
```

File: test/integration/buyers/accounts_controller_test.rb [hunk 5]
```diff
@@ -91,6 +91,42 @@ def test_show
...
+      post admin_buyers_account_invoices_path(buyer)
+      assert_response :forbidden
+
+      invoice = FactoryBot.create(:invoice, :buyer_account => buyer, :provider_account => provider)
+      Invoice.any_instance.stubs(:find).returns(invoice)
+      get edit_admin_buyers_account_invoice_path(buyer, invoice)
+      assert_response :forbidden
+    end
+
+    test 'memeber user with billing permission can manage invoice' do
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:110]: 👀 

```suggestion
    test 'member user with billing permission can manage invoice' do
```

File: test/integration/buyers/accounts_controller_test.rb [hunk 6]
```diff
@@ -91,6 +91,42 @@ def test_show
...
+      provider.settings.allow_finance!
+
+      get admin_buyers_account_invoices_path(buyer)
+      assert_response :forbidden
+
+      post admin_buyers_account_invoices_path(buyer)
+      assert_response :forbidden
+
+      invoice = FactoryBot.create(:invoice, :buyer_account => buyer, :provider_account => provider)
+      Invoice.any_instance.stubs(:find).returns(invoice)
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:105]: Why? 😕

File: test/integration/buyers/accounts_controller_test.rb [hunk 7]
```diff
@@ -91,6 +91,42 @@ def test_show
...
+      get admin_buyers_account_invoices_path(buyer)
+      assert_response :success
+
+      assert_difference ->{ Invoice.count }, 1 do
+      post admin_buyers_account_invoices_path(buyer)
+      assert_response :redirect
+      end
+
+      invoice = FactoryBot.create(:invoice, :buyer_account => buyer, :provider_account => provider)
+      Invoice.any_instance.stubs(:find).returns(invoice)
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:125]: Same as https://github.com/3scale/porta/pull/2047/files#r459260108

File: test/integration/buyers/accounts_controller_test.rb (hunk 5 above)
Comment (HemaHG) [reply test/integration/buyers/accounts_controller_test.rb:110]: my typo sorry

File: app/controllers/buyers/invoices_controller.rb (hunk 2 above)
Comment (hallelujah) [reply app/controllers/buyers/invoices_controller.rb]: @Martouta For show it makes sense no?

File: test/integration/buyers/accounts_controller_test.rb [hunk 8]
```diff
@@ -91,6 +91,40 @@ def test_show
...
+
+      get admin_buyers_account_invoices_path(buyer)
+      assert_response :success
+
+      assert_difference ->{ Invoice.count }, 1 do
+      post admin_buyers_account_invoices_path(buyer)
+      assert_response :redirect
+      end
+
+      invoice = FactoryBot.create(:invoice, :buyer_account => buyer, :provider_account => provider)
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:123]:
```suggestion
      invoice = FactoryBot.create(:invoice, buyer_account: buyer, provider_account: provider)
```

File: test/integration/buyers/accounts_controller_test.rb [hunk 9]
```diff
@@ -91,6 +91,40 @@ def test_show
...
+      buyer = FactoryBot.create(:buyer_account, provider_account: provider)
+      provider.settings.allow_finance!
+
+      get admin_buyers_account_invoices_path(buyer)
+      assert_response :success
+
+      assert_difference ->{ Invoice.count }, 1 do
+      post admin_buyers_account_invoices_path(buyer)
+      assert_response :redirect
+      end
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:121]:
```suggestion
      assert_difference ->{ Invoice.count }, 1 do
        post admin_buyers_account_invoices_path(buyer)
        assert_response :redirect
      end
```

File: test/integration/buyers/accounts_controller_test.rb [hunk 10]
```diff
@@ -91,6 +91,40 @@ def test_show
...
+
+      invoice = FactoryBot.create(:invoice, :buyer_account => buyer, :provider_account => provider)
+      get edit_admin_buyers_account_invoice_path(buyer, invoice)
+      assert_response :forbidden
+    end
+
+    test 'member user with billing permission can manage invoice' do
+      user.member_permission_ids = [:partners, :finance]
+      user.save!
+      buyer = FactoryBot.create(:buyer_account, provider_account: provider)
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:112]:
```suggestion
      buyer = FactoryBot.create(:simple_buyer, provider_account: provider)
```

For efficiency 😄

File: test/integration/buyers/accounts_controller_test.rb [hunk 11]
```diff
@@ -91,6 +91,40 @@ def test_show
...
+      provider.settings.allow_finance!
+
+      get admin_buyers_account_invoices_path(buyer)
+      assert_response :forbidden
+
+      post admin_buyers_account_invoices_path(buyer)
+      assert_response :forbidden
+
+      invoice = FactoryBot.create(:invoice, :buyer_account => buyer, :provider_account => provider)
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:104]:
```suggestion
      invoice = FactoryBot.create(:invoice, buyer_account:, provider_account: provider)
```

File: test/integration/buyers/accounts_controller_test.rb [hunk 12]
```diff
@@ -91,6 +91,40 @@ def test_show
...
 
+    test "member user without billing permission cannot manage invoice" do
+      buyer = FactoryBot.create(:buyer_account, provider_account: provider)
+      provider.settings.allow_finance!
+
+      get admin_buyers_account_invoices_path(buyer)
+      assert_response :forbidden
+
+      post admin_buyers_account_invoices_path(buyer)
+      assert_response :forbidden
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:102]:
```suggestion
      assert_no_difference ->{ Invoice.count } do
        post admin_buyers_account_invoices_path(buyer)
        assert_response :forbidden
      end
```

File: test/integration/buyers/accounts_controller_test.rb [hunk 13]
```diff
@@ -91,6 +91,40 @@ def test_show
       assert_match 'Alaska Application App', response.body
     end
 
+    test "member user without billing permission cannot manage invoice" do
+      buyer = FactoryBot.create(:buyer_account, provider_account: provider)
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:95]:
```suggestion
      buyer = FactoryBot.create(:simple_buyer, provider_account: provider)
```

For efficiency.

File: test/integration/buyers/accounts_controller_test.rb [hunk 14]
```diff
@@ -91,6 +91,40 @@ def test_show
       assert_match 'Alaska Application App', response.body
     end
 
+    test "member user without billing permission cannot manage invoice" do
```
Comment (Martouta) [inline test/integration/buyers/accounts_controller_test.rb:94]:
```suggestion
    test 'member user without billing permission cannot manage invoice' do
```

**Comment by HemaHG** [comment]
line items is present inside the invoice, if you are not able to access the invoice then we cannot add line items also, but now I am confused @Martouta

**Comment by Martouta** [comment]
@HemaHG What you are describing is the navigation inside the UI. However, it is too easy to skip the UI and to requests directly to our routes/controllers, so we have to secure our line items controller 😉 
Plus, I do not know if we also have any endpoints in our API related to line_items or invoices 🤔 We should check that too 😝

File: app/controllers/finance/api/invoices_controller.rb [hunk 15]
```diff
@@ -6,6 +6,8 @@ class Finance::Api::InvoicesController < Finance::Api::BaseController
 
   STATES = Hash.new {|_, k| k}.merge('cancel' => 'cancelled').freeze
 
+  before_action :authorize_finance
```
Comment (guicassolato) [inline app/controllers/finance/api/invoices_controller.rb:9]: Shouldn't we add to `Finance::Api::PaymentTransactionsController` too? Why not adding to `Finance::Api::BaseController` then?

File: app/controllers/buyers/invoices_controller.rb [hunk 16]
```diff
@@ -4,6 +4,7 @@ class Buyers::InvoicesController < Buyers::BaseController
 
   helper_method :allow_edit?
 
+  before_action :authorize_finance
```
Comment (Martouta) [inline app/controllers/buyers/invoices_controller.rb:7]: why not reusing this?:
```ruby
include Finance::ControllerRequirements
```

File: test/integration/finance/provider/dashboards_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+    login! @provider
+  end
+
+  test 'raise exception if finance switch is denied' do
+    assert @provider.settings.finance.denied?
+    get admin_finance_root_path
+    assert_response :forbidden
+  end
+
+  # TODO: add some fake invoices
```
Comment (hallelujah) [inline test/integration/finance/provider/dashboards_controller_test.rb:17]: What about this TODO?

**Pull request merged**